metadata-service/factories/src/main/java/com/linkedin/metadata/boot/UpgradeStep.java

```diff
@@ -0,0 +1,134 @@
+  }
+
+  @Nonnull
+  @Override
+  public ExecutionMode getExecutionMode() {
+    return ExecutionMode.ASYNC;
+  }
+
+  public abstract void upgrade() throws Exception;
+
+  private boolean hasUpgradeRan() {
+    try {
+      EntityResponse response = _entityService.getEntityV2(Constants.DATA_HUB_UPGRADE_ENTITY_NAME, upgradeUrn,
+          Collections.singleton(Constants.DATA_HUB_UPGRADE_REQUEST_ASPECT_NAME));
+
+      if (response != null && response.getAspects().containsKey(Constants.DATA_HUB_UPGRADE_REQUEST_ASPECT_NAME)) {
+        DataMap dataMap = response.getAspects().get(Constants.DATA_HUB_UPGRADE_REQUEST_ASPECT_NAME).getValue().data();
+        DataHubUpgradeRequest request = new DataHubUpgradeRequest(dataMap);
+        if (request.hasVersion() && request.getVersion().equals(version)) {
+          log.info(String.format("Upgrade %s has run before for version %s. Skipping...", upgradeId, version));
+          return true;
+        }
+      }
+    } catch (Exception e) {
+      log.error("Error when checking to see if datahubUpgrade entity exists. Commencing with upgrade...", e);
+      return false;
+    }
+    return false;
+  }
+
+  private void ingestUpgradeRequestAspect() throws URISyntaxException {
+    final AuditStamp auditStamp =
+        new AuditStamp().setActor(Urn.createFromString(Constants.SYSTEM_ACTOR)).setTime(System.currentTimeMillis());
+    final DataHubUpgradeRequest upgradeRequest =
+        new DataHubUpgradeRequest().setTimestampMs(System.currentTimeMillis()).setVersion(version);
+
+    final MetadataChangeProposal upgradeProposal = new MetadataChangeProposal();
+    upgradeProposal.setEntityUrn(upgradeUrn);
+    upgradeProposal.setEntityType(Constants.DATA_HUB_UPGRADE_ENTITY_NAME);
+    upgradeProposal.setAspectName(Constants.DATA_HUB_UPGRADE_REQUEST_ASPECT_NAME);
+    upgradeProposal.setAspect(GenericRecordUtils.serializeAspect(upgradeRequest));
+    upgradeProposal.setChangeType(ChangeType.UPSERT);
+
+    _entityService.ingestProposal(upgradeProposal, auditStamp);
+  }
+
+  private void ingestUpgradeResultAspect() throws URISyntaxException {
+    final AuditStamp auditStamp =
+        new AuditStamp().setActor(Urn.createFromString(Constants.SYSTEM_ACTOR)).setTime(System.currentTimeMillis());
+    final DataHubUpgradeResult upgradeResult = new DataHubUpgradeResult().setTimestampMs(System.currentTimeMillis());
+
+    final MetadataChangeProposal upgradeProposal = new MetadataChangeProposal();
+    upgradeProposal.setEntityUrn(upgradeUrn);
+    upgradeProposal.setEntityType(Constants.DATA_HUB_UPGRADE_ENTITY_NAME);
+    upgradeProposal.setAspectName(Constants.DATA_HUB_UPGRADE_RESULT_ASPECT_NAME);
+    upgradeProposal.setAspect(GenericRecordUtils.serializeAspect(upgradeResult));
+    upgradeProposal.setChangeType(ChangeType.UPSERT);
+
+    _entityService.ingestProposal(upgradeProposal, auditStamp);
+  }
+
+  private void cleanUpgradeAfterError() {
+    _entityService.deleteUrn(upgradeUrn);
```

Comment:
What if this operation fails? It would be nice to make sure if this throws, we can at least log an error saying "Super bad thing just happened - we've locked ourselves from performing the upgrade" so we know that manual invervention will be required to solve this

Reply:
cleaned this up a bit so the error message is logged in this function